tests/bot/cogs/sync/test_base.py

```diff
@@ -349,7 +339,10 @@ def setUp(self):
                 if ctx is not None:
                     ctx.send.return_value = message
 
-                self.syncer._get_confirmation_result = helpers.AsyncMock(return_value=(False, None))
+                diff = _Diff({1, 2, 3}, {4, 5}, None)
+                self.syncer._get_diff.return_value = diff
```

Comment:
Why was this added?

Reply:
I should probably just set an explicit `MagicMock` as the return value instead of an actual `_Diff` object, but the "issue" here is that the `AsyncMock` you've used to mock `_get_diff` will return another `AsyncMock` as its child when its called and awaited. (The default behavior for mocks is to return a child mock of its own type.)

Because of that "child is like parent" behavior, this will happen in [lines 194-196 of bot.cogs.sync.syncers](https://github.com/python-discord/bot/blob/c81a4d401ea434e98b0a1ece51d3d10f1a3ad226/bot/cogs/sync/syncers.py#L194-L196):

```py
        # `diff` will be a child AsyncMock as "child is like parent"
        diff = await self._get_diff(guild)

        # `diff._asdict` will be an AsyncMock as "child is like parent" and,
        # after being called, it will return an unawaited coroutine object
        # we assign the name `diff_dict`:
        diff_dict = diff._asdict()

        # `diff_dict` is still an unawaited coroutine object meaning that it
        # doesn't have an `items()` method:
        totals = {k: len(v) for k, v in diff_dict.items() if v is not None}
```

This will lead to the following exception when running the tests:

```py
======================================================================
ERROR: test_sync_confirmation_context_redirect (tests.bot.cogs.sync.test_base.SyncerSyncTests) (ctx=None, author=<MockMember name='mock.user' spec_set='Member' id='140583452034864'>, message=None)
If ctx is given, a new message should be sent and author should be ctx's author.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/sebastiaan/pydis/repositories/bot/tests/bot/cogs/sync/test_base.py", line 348, in test_sync_confirmation_context_redirect
    await self.syncer.sync(guild, ctx)
  File "/home/sebastiaan/pydis/repositories/bot/bot/cogs/sync/syncers.py", line 196, in sync
    totals = {k: len(v) for k, v in diff_dict.items() if v is not None}
AttributeError: 'coroutine' object has no attribute 'items'
```

The solution is to assign an explicit return value so the parent mock doesn't "guess" and return an object of its own type. However, instead of assigning a specific `_Diff` object as the return value, I should probably have gone with a `MagicMock` to signify that it's not an important return value; it's just something that needs to support/mimic the API we use on it.

Reply:
Oh yeah I remember you mentioning that weird issue in VC. However, I ran it without that change locally and the test passed.

Reply:
Actually, that may be because I changed the syncer to use autospecs (which I believe is smarter about specing attributes than just using the same mock as the parent).

Reply:
That's for the `test_cog` file, right? The `test_base` file uses this:

```py
class TestSyncer(Syncer):
    """Syncer subclass with mocks for abstract methods for testing purposes."""

    name = "test"
    _get_diff = mock.AsyncMock()
    _sync = mock.AsyncMock()
```

If I comment out line 343 in `test_base`, I do get the mocking behavior above (and the exception/unawaited coroutine warnings).

Reply:
Yeah you're right about the autospecs. That change is actually irrelavant. Anyway, I ran the test again and it's failing now. I don't know why it appeared to pass last night.